Documentation/Browser/Patterns.md

```diff
@@ -0,0 +1,130 @@
+`TRY_OR_THROW_OOM` macro). This avoids a situation where almost everything returns a generic
+`WebIDL::ExceptionOr<T>`.
+
+### `Web::WebIDL::ExceptionOr<T>`
+
+This is the most common and at the same time most broad error type in LibWeb. Internally it stores a
+variant of supported errors:
+
+-   `SimpleException`
+-   `JS::NonnullGCPtr<DOMException>`
+-   `JS::Completion` (from `JS::ThrowCompletionOr<T>`, assumed to be of `Type::Throw`)
+
+Use this error type for anything that needs to interact with the JS bindings, which will generally
+know how to turn any of the internally supported errors into JS objects.
+
+### `Web::WebIDL::SimpleException`
+
+This is a thin wrapper around various built-in errors from ECMAScript:
+
+-   `EvalError`
+-   `RangeError`
+-   `ReferenceError`
+-   `TypeError`
+-   `URIError`
+
+Instead of constructing one of these directly, create a `SimpleException` with the appropriate type
+and message instead whenever required by a web spec. These will be converted into actual JS objects
+in the bindings layer.
+
+### `Web::WebIDL::DOMException`
+
+This is an error type from the WebIDL spec specifically for web AOs where none of the JS built-in
+error types are sufficient. Like with `SimpleException`, use when indicated by the spec.
+
+### `JS::ThrowCompletionOr<T>`
+
+This is an error type from LibJS, which uses "completions" to propagate errors and other types of AO
+results. Don't use this in LibWeb unless absolutely necessary, e.g. when overriding a `JS::Object`
+virtual method that returns this type.
+
+At the call site, these should be wrapped in a `ExceptionOr<T>` as soon as possible for further
+propagation.
+
+## Comments
+
+As in LibJS, **all** functions that represent an operation or JS function from a web specification
+must have:
+
+-   A spec link, above the function definition:
+
+    ```cpp
+    // https://fetch.spec.whatwg.org/#concept-fetch
+    WebIDL::ExceptionOr<JS::NonnullGCPtr<Infrastructure::FetchController>> fetch(JS::Realm& realm, Infrastructure::Request& request, Infrastructure::FetchAlgorithms const& algorithms, UseParallelQueue use_parallel_queue)
+    {
+        // ...
+    }
+    ```
+
+-   Comments for each individual step of the operation:
+
+    ```cpp
+    // 1. Assert: request’s mode is "navigate" or processEarlyHintsResponse is null.
+    VERIFY(request.mode() == Infrastructure::Request::Mode::Navigate || !algorithms.process_early_hints_response().has_value());
+
+    // 2. Let taskDestination be null.
+    JS::GCPtr<JS::Object> task_destination;
+
+    // ...
+    ```
+
+    -   If a step cannot be implemented at the time, prepend it with a `FIXME`
+    -   Add a blank line between code and the next comment
+
+-   Optimizations (e.g. fast paths) should be marked as such with an `// OPTIMIZATION:` comment
+    explaining the reasoning
+
+-   All functionality that is not directly implementing a concept from a spec should be clearly
+    marked as such
```

Comment:
This seems overly strong, given that there's tons of code in LibWeb (layout, painting, hit testing, etc) that doesn't implement a spec (because there isn't one that covers the particular area.)

Reply:
yeah that's fair, seems like I'm spending too much time in the well-written JS-y areas of those web specs 😄
Relaxed this a little bit.